service/service.go

```diff
@@ -2,90 +2,161 @@ package service
-	router.StrictSlash(true).Path("/health").HandlerFunc(hc.Handler)
-	hc.Start(ctx)
+	// Get HTTP router and server with middleware
+	router := mux.NewRouter()
+	middle := svc.createMiddleware(svc.Config)
+	svc.Server = svc.ServiceList.GetHTTPServer(svc.Config.BindAddr, middle.Then(router))
+
+	// Setup the API
+	topicPermissions, permissions := getAuthorisationHandlers(ctx, svc.Config)
+	svc.API = api.Setup(ctx, svc.Config, router, store, topicPermissions, permissions)
+
+	svc.HealthCheck.Start(ctx)
 
 	// Run the http server in a new go-routine
 	go func() {
-		if err := srv.ListenAndServe(); err != nil {
+		if err := svc.Server.ListenAndServe(); err != nil {
 			svcErrors <- errors.Wrap(err, "failure in http listen and serve")
 		}
 	}()
 
-	return &Service{
-		Config:      cfg,
-		Router:      router,
-		API:         a,
-		HealthCheck: hc,
-		ServiceList: serviceList,
-		Server:      srv,
-		MongoDB:     mongoDB,
-	}, nil
+	return nil
+}
+
+func getAuthorisationHandlers(ctx context.Context, cfg *config.Configuration) (api.AuthHandler, api.AuthHandler) {
+	if cfg.EnablePermissionsAuth == false {
+		log.Event(ctx, "feature flag not enabled defaulting to nop authZ impl", log.INFO, log.Data{"feature": "ENABLE_PERMISSIONS_AUTHZ"})
+		return &auth.NopHandler{}, &auth.NopHandler{}
+	}
+
+	log.Event(ctx, "feature flag enabled", log.INFO, log.Data{"feature": "ENABLE_PERMISSIONS_AUTHZ"})
+
+	authClient := auth.NewPermissionsClient(dphttp.NewClient())
+	authVerifier := auth.DefaultPermissionsVerifier()
+
+	// for checking caller permissions when we have a datasetID, collection ID and user/service token
+	datasetPermissions := auth.NewHandler(
+		auth.NewDatasetPermissionsRequestBuilder(cfg.ZebedeeURL, "dataset_id", mux.Vars),
+		authClient,
+		authVerifier,
+	)
+
+	// for checking caller permissions when we only have a user/service token
+	permissions := auth.NewHandler(
+		auth.NewPermissionsRequestBuilder(cfg.ZebedeeURL),
+		authClient,
+		authVerifier,
+	)
+
+	return datasetPermissions, permissions
+}
+
+// CreateMiddleware creates an Alice middleware chain of handlers
+// to forward collectionID from cookie from header
+func (svc *Service) createMiddleware(cfg *config.Configuration) alice.Chain {
+
+	// healthcheck
+	healthcheckHandler := newMiddleware(svc.HealthCheck.Handler, "/health")
+	middleware := alice.New(healthcheckHandler)
+
+	// Only add the identity middleware when running in publishing.
+	if cfg.EnablePrivateEndpoints {
+		middleware = middleware.Append(dphandlers.IdentityWithHTTPClient(svc.IdentityClient))
+	}
+
+	// collection ID
+	middleware = middleware.Append(dphandlers.CheckHeader(dphandlers.CollectionID)) //!!! is this appropriate for topic-api ? ... ask Eleanor and David Subiros
+
+	return middleware
+}
+
+// newMiddleware creates a new http.Handler to intercept /health requests.
+func newMiddleware(healthcheckHandler func(http.ResponseWriter, *http.Request), path string) func(http.Handler) http.Handler {
```

Comment:
(just an observation for the future): I think that this middleware should be part of the health check library so that we don't need to have it in all the repos.

Reply:
OK.
Thanks.